docs/architecture/adr-033-protobuf-inter-module-comm.md

```diff
@@ -0,0 +1,341 @@
+private member variables could be manipulated through reflection. Golang does not support reflection on a function
+closure's captured variables and direct manipulation of memory would be needed for a truly malicious module to bypass
+the `ModuleKey` security.
+
+The two `ModuleKey` types are `RootModuleKey` and `DerivedModuleKey`:
+
+```go
+func Invoker(callInfo CallInfo) func(ctx context.Context, request, response interface{}, opts ...interface{}) error
+
+type CallInfo {
+  Method string
+  Caller ModuleID
+}
+
+type RootModuleKey struct {
+  moduleName string
+  invoker Invoker
+}
+
+type DerivedModuleKey struct {
+  moduleName string
+  path []byte
+  invoker Invoker
+}
+```
+
+A module can get access to a `DerivedModuleKey`, using the `Derive(path []byte)` method on `RootModuleKey` and then
+would use this key to authenticate `Msg`s from a sub-account. Ex:
+
+```go
+package foo
+
+func (fooMsgServer *MsgServer) Bar(ctx context.Context, req *MsgBar) (*MsgBarResponse, error) {
+  derivedKey := fooMsgServer.moduleKey.Derive(req.SomePath)
+  bankMsgClient := bank.NewMsgClient(derivedKey)
+  res, err := bankMsgClient.Balance(ctx, &bank.MsgSend{FromAddress: derivedKey.Address(), ...})
+  ...
+}
+```
+
+In this way, a module can gain permissioned access to a root account and any number of sub-accounts and send
+authenticated `Msg`s from these accounts. The `Invoker` `callInfo.Caller` parameter is used under the hood to
+distinguish between different module accounts, but either way the function returned by `Invoker` only allows `Msg`s
+from either the root or a derived module account to pass through.
+
+Note that `Invoker` itself returns a function closure based on the `CallInfo` passed in. This will allow client implementations
+in the future that cache the invoke function for each method type avoiding the overhead of hash table lookup.
+This would reduce the performance overhead of this inter-module communication method to the bare minimum required for
+checking permissions.
+
+Below is a rough sketch of the implementation of `grpc.ClientConn.Invoke` for `RootModuleKey`:
+
+```go
+func (key RootModuleKey) Invoke(ctx context.Context, method string, args, reply interface{}, opts ...grpc.CallOption) error {
+  f := key.invoker(CallInfo {Method: method, Caller: ModuleID {ModuleName: key.moduleName}})
+  return f(ctx, args, reply)
+}
+```
+
+### `AppModule` Wiring and Requirements
+
+In [ADR 031](./adr-031-msg-service.md), the `AppModule.RegisterService(Configurator)` method was introduced. To support
+inter-module communication, we extend the `Configurator` interface to pass in the `ModuleKey` and to allow modules to
+specify their dependencies on other modules using `RequireServer()`:
+
+
+```go
+type Configurator interface {
+   MsgServer() grpc.Server
+   QueryServer() grpc.Server
+
+   ModuleKey() ModuleKey
+   RequireServer(serverInterface interface{})
+}
```

Comment:
If configurator is kind of a config, then it will be more natural to use a `struct` here, unless we require a Configurator to have certain methods and be able to mock them in a more advanced way. Basically I don't see why we prefer interface here - for simple mocking we can just set expected structure values.

Reply:
We are mocking it for testing in regen already. I think it's useful.